### PR TITLE
ddl_notifier: add subscription API

### PR DESCRIPTION
**Pull request by lance6716**
### What problem does this PR solve?

Issue Number: ref #55722

Problem Summary:

### What changed and how does it work?

Add RegisterHandler, etc for subscribers according to RFC

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/ddl/notifier/subscribe.go [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+	ctx = kv.WithInternalSourceType(ctx, kv.InternalDDLNotifier)
+	ticker := time.NewTicker(n.pollInterval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			if err := n.processEvents(ctx); err != nil {
+				log.FromContext(ctx).Error("Error processing events", zap.Error(err))
```
Comment (fzzf678) [inline pkg/ddl/notifier/subscribe.go:128]: This logger seems like the `DDLIngestLogger`, here `DDLLogger`  looks more accurate
```
[2024/09/20 16:52:54.168 +08:00] [ERROR] [subscribe.go:151] ["Error processing event"] **[category=ddl-ingest]** [ddlJobID=3] [multiSchemaChangeSeq=-1] [handlerID=0] [error=EOF] [errorVerbose="EOF\ngithub.com/pingcap/errors.AddStack\n\t/Users/fanzhou/.gvm/pkgsets/go1.22.4/global/pkg/mod/github.com/pingcap/errors@v0.11.5-0.20240318064555-6bd07397691f/errors.go:178\ngithub.com/pingcap/errors.Trace\n\t/Users/fanzhou/.gvm/pkgsets/go1.22.4/global/pkg/mod/github.com/pingcap/errors@v0.11.5-0.20240318064555-6bd07397691f/juju_adaptor.go:15\ngithub.com/pingcap/tidb/pkg/ddl/notifier.(*ddlNotifier).processEventForHandler\n\t/Users/fanzhou/Documents/GitHub/tidb/pkg/ddl/notifier/subscribe.go:191\ngithub.com/pingcap/tidb/pkg/ddl/notifier.(*ddlNotifier).processEvents\n\t/Users/fanzhou/Documents/GitHub/tidb/pkg/ddl/notifier/subscribe.go:147\ngithub.com/pingcap/tidb/pkg/ddl/notifier.(*ddlNotifier).Start\n\t/Users/fanzhou/Documents/GitHub/tidb/pkg/ddl/notifier/subscribe.go:126\ngithub.com/pingcap/tidb/pkg/ddl/notifier.StartDDLNotifier\n\t/Users/fanzhou/Documents/GitHub/tidb/pkg/ddl/notifier/subscribe.go:113\ngithub.com/pingcap/tidb/pkg/ddl/notifier_test.TestBasicPubSub.func2\n\t/Users/fanzhou/Documents/GitHub/tidb/pkg/ddl/notifier/testkit_test.go:105\nruntime.goexit\n\t/Users/fanzhou/.gvm/gos/go1.22.4/src/runtime/asm_arm64.s:1222"]

```

File: pkg/ddl/notifier/subscribe.go [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+	for _, event := range events {
+		for handlerID, handler := range n.handlers {
+			if _, ok := skipHandlers[handlerID]; ok {
+				continue
+			}
+			if err2 := n.processEventForHandler(ctx, event, handlerID, handler); err2 != nil {
+				skipHandlers[handlerID] = struct{}{}
+
+				if !goerr.Is(err2, ErrNotReadyRetryLater) {
+					log.FromContext(ctx).Error("Error processing event",
```
Comment (fzzf678) [inline pkg/ddl/notifier/subscribe.go:152]: ditto

File: pkg/ddl/notifier/subscribe.go (hunk 1 above)
Comment (lance6716) [reply pkg/ddl/notifier/subscribe.go]: https://github.com/pingcap/tidb/pull/56182/commits/734a9c580fca6bb8dfba25dbd8316a19d2b1e5ed

**Review by fzzf678 — APPROVED** [review]

**Comment by lance6716** [comment]
ptal @Rustin170506 @time-and-fate

File: pkg/ddl/notifier/events.go [hunk 3]
```diff
@@ -423,7 +423,7 @@ func (s *SchemaChangeEvent) MarshalJSON() ([]byte, error) {
 func (s *SchemaChangeEvent) UnmarshalJSON(b []byte) error {
 	var j jsonSchemaChangeEvent
 	err := json.Unmarshal(b, &j)
-	if err != nil {
```
Comment (Rustin170506) [inline pkg/ddl/notifier/events.go:426]: Do we need to add a test to cover it?

File: pkg/ddl/notifier/subscribe.go [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+var ErrNotReadyRetryLater = errors.New("not ready, retry later")
+
+// HandlerID is the type of the persistent ID used to register a handler. Every
+// ID occupies a bit in a BIGINT column, so at most we can only have 64 IDs. To
+// avoid duplicate IDs, all IDs should be defined in below declaration.
+type HandlerID int
+
+const (
+	// TestHandlerID is used for testing only.
+	TestHandlerID HandlerID = iota
```
Comment (Rustin170506) [inline pkg/ddl/notifier/subscribe.go:63]: I guess it can be a fixed number instead of an `iota`?

File: pkg/ddl/notifier/subscribe.go [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+)
+
+// RegisterHandler must be called with an exclusive and fixed HandlerID for each
+// handler to register the handler. Illegal ID will panic. RegisterHandler should
+// not be called after the global ddlNotifier is started.
+//
+// RegisterHandler is not concurrency-safe.
+func RegisterHandler(id HandlerID, handler SchemaChangeHandler) {
+	intID := int(id)
+	if intID < 0 || intID >= 64 {
```
Comment (Rustin170506) [inline pkg/ddl/notifier/subscribe.go:73]: Would it be a good idea to turn these magic numbers into constants?

File: pkg/ddl/notifier/subscribe.go [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+	// we should ensure deliver order of events to a handler, so if a handler returns
+	// error for previous events it should not receive later events.
+	skipHandlers := make(map[HandlerID]struct{})
+	for _, event := range events {
+		for handlerID, handler := range n.handlers {
+			if _, ok := skipHandlers[handlerID]; ok {
+				continue
+			}
+			if err2 := n.processEventForHandler(ctx, event, handlerID, handler); err2 != nil {
+				skipHandlers[handlerID] = struct{}{}
```
Comment (Rustin170506) [inline pkg/ddl/notifier/subscribe.go:149]: Do you mean that if one event handling fails, we should skip all events for this handler?

Comment (Rustin170506) [reply pkg/ddl/notifier/subscribe.go:149]: I guess it makes sense for 99% of handlers.

File: pkg/ddl/notifier/subscribe.go [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+		case <-ticker.C:
+			if err := n.processEvents(ctx); err != nil {
+				logutil.Logger(ctx).Error("Error processing events", zap.Error(err))
+			}
+		}
+	}
+}
+
+func (n *ddlNotifier) processEvents(ctx context.Context) error {
+	events, err := n.store.List(ctx, sess.NewSession(n.ownedSCtx))
```
Comment (Rustin170506) [inline pkg/ddl/notifier/subscribe.go:135]: Maybe we should call it `changes`?

**Review by Rustin170506 — COMMENTED** [review]
Thanks!

File: pkg/ddl/notifier/subscribe.go (hunk 5 above)
Comment (lance6716) [reply pkg/ddl/notifier/subscribe.go:73]: I can add comments here. They do not have a meaningful name for the constant

File: pkg/ddl/notifier/subscribe.go (hunk 6 above)
Comment (lance6716) [reply pkg/ddl/notifier/subscribe.go:149]: We must do this because the notifier should guarantee the delivery order.

**Comment by lance6716** [comment]
ping @Rustin170506 @time-and-fate @winoros

File: pkg/ddl/notifier/subscribe.go (hunk 6 above)
Comment (Rustin170506) [reply pkg/ddl/notifier/subscribe.go:151]: oh, I forgot that.

Comment (Rustin170506) [reply pkg/ddl/notifier/subscribe.go:151]: Maybe we can add some comments here?

Comment (lance6716) [reply pkg/ddl/notifier/subscribe.go:151]: I have added in line 142

**Review by Rustin170506 — APPROVED** [review]
Looks good to me. Thanks! 👍

File: pkg/ddl/notifier/subscribe.go [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+	}
+	defer func() {
+		if err == nil {
+			err = errors.Trace(se.Commit(ctx))
+		} else {
+			se.Rollback()
+		}
+	}()
+
+	// TODO: Should we attach a timeout to this ctx?
```
Comment (time-and-fate) [inline pkg/ddl/notifier/subscribe.go:192]: Another idea is to record the time spent and print a log when it exceeds a threshold.

**Review by winoros — APPROVED** [review]

**Pull request merged**